docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,242 @@
+{ "parent": "<parent-device-id>" }
+```
+
+The `parent-device-id` is the device-id of the direct parent that the child device is connected to.
+The payload can have other fields describing the capabilities of that device as well (config management, software management etc).
+
+```admonish warning
+The exact topic keys and payload format for this init contract can be discussed and refined separately.
+The focus here is just on the MQTT topic structure.
+```
+
+Thin-edge needs to maintain the lineage (hierarchy of parent devices) of all descendent child devices in its internal state,
+so that it can be looked up while receiving any data from them.
+Even though the child device only declares its immediate parent in the registration message,
+the entire lineage can be traced back with a recursive lookup on that `parent-device-id`.
+
+The registration status, whether the device registration succeeded or not, is sent back on `tedge/main/init/child/res/<child-id>`,
+with the internal device id used by thin-edge to uniquely identify that device as follows:
+
+```json
+{
+   "id": "<internal-id>",
+   "status": "successful" 
+}
+```
+
+A failure is indicated with a failed status: `{ "status": "failed" }`.
+
+Similarly, a service can register itself by sending the init message to `tedge/main/init/service/req/<service-id>`
+and expect the response on `tedge/main/init/service/res/<service-id>`.
+
+Once the registration is complete, these nested child devices can use the `tedge/descendent/<child-id>` topic prefix
+to send telemetry data or receive commands as follows:
+
+* Measurement: `tedge/descendent/<internal-id>/telemetry/measurements`
+* Command: `tedge/descendent/<internal-id>/commands/req/software_list`
+
+#### Automatic registration
+
+Even though an explicit registration is always desired,
+automatic registration is also supported for services and immediate child devices of the thin-edge device,
+on the receipt of the very first message from them.
+But, this automatic registration is not supported for descendent child devices.
+An explicit registration is mandatory for them.
+
+**Pros**
+
+1. The context on whether the data is coming from the parent, a child or a service is clear from the topics.
+1. Automatic registration is possible at least for child devices and main device services.
+
+**Cons**
+
+1. No support for services on child devices.
+   Even though the support can be added, it would make the topics extremely long.
+   E.g: `tedge/child/<child-id>/service/<service-id>/telemetry/measurements`
+1. The topics are fairly long and with extensions might easily cross the 7 sub-topic limit fo AWS 
+
+### Unified topics for every "thing"
+
+Use just `id`s in the topic for everything: including parent device, child devices and services.
+with just the distinction between device or service as follows:
+`tedge/<context>/<id>/...`
+
+For device(parent or child): `tedge/device/<id>/...`
+For services(on parent or child): `tedge/service/<id>/...`
```

Comment:
Some more additional context can be added with this variation:

```
For tedge device: `tedge/main-device/<id>/...`
For child device: `tedge/child-device/<id>/...`
For descendent device: `tedge/descendent-device/<id>/...`
For services on tedge device: `tedge/main-service/<id>/...`
For services on descendent device: `tedge/descendent-service/<id>/...`
```

...sacrificing the benefit of having unified topics.

The other extreme is to remove the "device vs service" context as well, by just keeping the `id` in the topics, as in `tedge/<id>/...` saving one more subtopic. But I don't see much benefits with that extreme.

Reply:
The difference in processing is mostly in the registration step, and if we support lazy instantiation (e.g. create a device/service on the first publish of a measurement) then the context is very important because devices/child-devices/services all differ in how they are registered.

Reply:
Another important aspect is **observability** as well. It'd be far more easier for us to interpret logs or debug issues when these messages have some context also associated with it, as a message trace on `tedge/#` is usually the main first that we look at, for any debugging.

It also helps in easier filtering of messages as well. Yes, even with ids, we can do a lookup first and then apply the filtering. But having the context in the messages makes it a lot more easier.